modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/HeapLockManager.java

```diff
@@ -328,10 +349,359 @@ private LockState adjustLockState(LockState state, LockState v) {
+                }
+
+                IgniteBiTuple<Lock, CompletableFuture<Lock>> tuple = slockWaiters.get(txId);
+
+                if (tuple != null) {
+                    return tuple.get1();
+                }
+            } finally {
+                stripedLock.readLock(idx).unlock();
+            }
+
+            return null;
+        }
+
+        @Override
+        public boolean coarse() {
+            return true;
+        }
+
+        /**
+         * Acquires a lock.
+         *
+         * @param txId Tx id.
+         * @param lockKey Lock key.
+         * @param lockMode Lock mode.
+         * @return The future.
+         */
+        public CompletableFuture<Lock> acquire(UUID txId, LockKey lockKey, LockMode lockMode) {
+            switch (lockMode) {
+                case S:
+                    stripedLock.writeLock().lock();
+
+                    try {
+                        // IX-locks can't be modified under the striped write lock.
+                        if (!ixlockOwners.isEmpty()) {
+                            if (ixlockOwners.containsKey(txId)) {
+                                if (ixlockOwners.size() == 1) {
+                                    // Safe to upgrade.
+                                    track(txId, this); // Double track.
+                                    Lock lock = new Lock(lockKey, lockMode, txId);
+                                    slockOwners.putIfAbsent(txId, lock);
+                                    return completedFuture(lock);
+                                } else {
+                                    // Attempt to upgrade to SIX in the presence of concurrent transactions. Deny lock attempt.
+                                    for (Lock lock : ixlockOwners.values()) {
+                                        if (!lock.txId().equals(txId)) {
+                                            return failedFuture(lockException(txId, lock.txId()));
+                                        }
+                                    }
+                                }
+
+                                assert false : "Should not reach here";
+                            }
+
+                            UUID holderTx = ixlockOwners.keySet().iterator().next();
+                            CompletableFuture<Void> res = fireEvent(LOCK_CONFLICT, new LockEventParameters(txId, holderTx));
+                            // TODO: https://issues.apache.org/jira/browse/IGNITE-21153
+                            if (res.isCompletedExceptionally()) {
+                                return failedFuture(abandonedLockException(txId, holderTx));
+                            }
+
+                            // Validate reordering with IX locks.
+                            for (Lock lock : ixlockOwners.values()) {
+                                // Allow only high priority transactions to wait.
+                                if (TxIdPriorityComparator.INSTANCE.compare(lock.txId(), txId) < 0) {
```

Comment:
TxIdComparator from deadlockPreventionPolicy should be used here, otherwise we will have lock conflicts if we change the tx id ordering in the deadlock prevention policy.

Reply:
🆗